src/query/src/range_select/plan_rewrite.rs

```diff
@@ -108,34 +112,77 @@ fn parse_expr_to_string(args: &[Expr], i: usize) -> DFResult<String> {
-            Ok(interval_to_duration(Interval::from_i32(*i)))
+        Some(expr) => {
+            let ms = evaluate_expr_to_millisecond(args, i)?;
+            if ms <= 0 {
+                return Err(dispose_parse_error(Some(expr)));
+            }
+            Ok(Duration::from_millis(ms as u64))
+        }
+        None => Err(dispose_parse_error(None)),
+    }
+}
+
+/// Evaluate a time calculation expr, case like:
+/// 1. `INTERVAL '1' day + INTERVAL '1 year 2 hours 3 minutes'`
+/// 2. `now() - INTERVAL '1' day`
+///
+/// Output a millisecond timestamp
+fn evaluate_expr_to_millisecond(args: &[Expr], i: usize) -> DFResult<i64> {
+    let Some(expr) = args.get(i) else {
+        return Err(dispose_parse_error(None));
+    };
+    let execution_props = ExecutionProps::new().with_query_execution_start_time(Utc::now());
+    let info = SimplifyContext::new(&execution_props).with_schema(Arc::new(DFSchema::empty()));
+    let interval_to_ms =
+        |interval: Interval| -> i64 { (interval.to_nanosecond() / NANOS_PER_MILLI as i128) as i64 };
+    let simplify_expr = ExprSimplifier::new(info).simplify(expr.clone())?;
+    match simplify_expr {
+        Expr::Literal(ScalarValue::TimestampNanosecond(ts_nanos, _))
+        | Expr::Literal(ScalarValue::DurationNanosecond(ts_nanos)) => {
+            ts_nanos.map(|v| v / 1_000_000)
+        }
+        Expr::Literal(ScalarValue::TimestampMicrosecond(ts_micros, _))
+        | Expr::Literal(ScalarValue::DurationMicrosecond(ts_micros)) => {
+            ts_micros.map(|v| v / 1_000)
         }
-        Some(Expr::Literal(ScalarValue::IntervalDayTime(Some(i)))) => {
-            Ok(interval_to_duration(Interval::from_i64(*i)))
+        Expr::Literal(ScalarValue::TimestampMillisecond(ts_millis, _))
+        | Expr::Literal(ScalarValue::DurationMillisecond(ts_millis)) => ts_millis,
+        Expr::Literal(ScalarValue::TimestampSecond(ts_secs, _))
+        | Expr::Literal(ScalarValue::DurationSecond(ts_secs)) => ts_secs.map(|v| v * 1_000),
+        Expr::Literal(ScalarValue::IntervalYearMonth(interval)) => {
+            interval.map(|v| interval_to_ms(Interval::from_i32(v)))
         }
-        Some(Expr::Literal(ScalarValue::IntervalMonthDayNano(Some(i)))) => {
-            Ok(interval_to_duration(Interval::from_i128(*i)))
+        Expr::Literal(ScalarValue::IntervalDayTime(interval)) => {
+            interval.map(|v| interval_to_ms(Interval::from_i64(v)))
         }
-        other => Err(dispose_parse_error(other)),
+        Expr::Literal(ScalarValue::IntervalMonthDayNano(interval)) => {
+            interval.map(|v| interval_to_ms(Interval::from_i128(v)))
+        }
+        _ => None,
     }
+    .ok_or(DataFusionError::Plan(format!(
+        "{} is not a expr can be evaluate and use in range query",
+        expr.display_name().unwrap_or_default()
+    )))
 }
 
 /// Parse the `align to` clause and return a UTC timestamp with unit of millisecond,
```

Comment:
out-of-topic: Will it be a problem if we want to process timestamps in other resolutions? e.g. nanosecond

Reply:
Nanoseconds are generally not processed. The original design goal should be to provide functions similar to PromQL range queries, so the syntax for specifying time durations also comes from PromQL, which the highest precision is milliseconds. In the execution of range plan, if the underlying time is nanoseconds, it will also be converted to milliseconds for calculation. There may indeed be potential resolution issues

Reply:
Does our document cover this point?

Reply:
Not show in document. The calculation is compatible with nanosecond timestamp data, but the underlying calculation uses milliseconds as the unit.

Reply:
Some parameters of the range query only support to millisecond precision, and the other part can specify a timestamp with any precision, but all the underlying calculation is based on millisecond precision.

Reply:
What if users would like to align to a microsecond or nanosecond timestamp? We allow users to pass timestamps in any resolution in SQL but the output may not as what users expect. It would be confusing without a document.

Reply:
I will add some explanation to document